subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt

```diff
@@ -57,49 +61,83 @@ interface EncryptionService : EncryptionConfiguration {
-            try {
-                keySource!!.getKey().also {
-                    val keyLength = it.encoded.size
-                    if (keyLength < 16) {
-                        throw InvalidKeyException("Encryption key length is $keyLength bytes, but must be at least 16 bytes long")
+        when {
+            // encryption is on by default
+            startParameter.encryptionRequested -> {
+                secretKeySourceFrom(startParameter).let { keySource ->
+                    try {
+                        secretKeyFrom(keySource, cacheBuilderFactory).also { key ->
+                            assertKeyLength(key)
+                            // acquire a cipher just to validate the key
+                            encryptionAlgorithm.newSession(key).encryptingCipher {}
+                        }
+                    } catch (e: EncryptionException) {
+                        if (e.message != null) {
+                            throw e
+                        }
+                        throw EncryptionException("Error loading encryption key from ${keySource.sourceDescription}", e.cause)
+                    } catch (e: Exception) {
+                        throw EncryptionException("Error loading encryption key from ${keySource.sourceDescription}", e)
                     }
-                    // acquire a cipher just to validate the key
-                    encryptionAlgorithm.newSession(it).encryptingCipher {}
                 }
-            } catch (e: EncryptionException) {
-                logger.warn("Encryption was requested but could not be enabled: ${e.message}")
-                null
-            } catch (e: Exception) {
-                logger.warn("Encryption was requested but could not be enabled: $e")
+            }
+
+            else -> {
                 null
             }
+        }
     }
 
-    override
-    val encryptionAlgorithm: EncryptionAlgorithm by lazy {
-        SupportedEncryptionAlgorithm.byTransformation(startParameter.encryptionAlgorithm)
+    private
+    fun assertKeyLength(key: SecretKey) {
+        val keyLength = key.encoded.size
+        if (keyLength < 16) {
+            throw InvalidKeyException("Encryption key length is $keyLength bytes, but must be at least 16 bytes long")
+        }
     }
 
     private
-    val keySource: SecretKeySource? by lazy {
-        if (encryptingRequested)
-            sourceKeyFromEnvironment()
-        else
-            null
+    fun secretKeyFrom(keySource: KeyStoreKeySource, cacheBuilderFactory: GlobalScopedCacheBuilderFactory) =
+        cacheBuilderFactory
+            .run { keySource.customKeyStoreDir?.let { createCacheBuilderFactory(it) } ?: this }
+            .createCacheBuilder("cc-keystore")
+            .withDisplayName("Gradle Configuration Cache keystore")
+            .withInitializer {
+                keySource.createKeyStoreAndGenerateKey(keyStoreFile())
+            }
+            .open().useToRun {
+                try {
```

Comment:
both `assert keystoreFile().file` and `assert keystoreFile().length() > 0` should always pass here, right?

Reply:
Right, unless the cache has been tampered with.